### ISSUE TITLE
Create an app for entering evidence 

### ISSUE THREAD

**Issue by prei007** [prei007/formative-assessment#4]
A reasonable first attempt (thanks to chatGPT):

```
library(shiny)
library(gRain)

# Define the Bayesian network (Chest Clinic / Asia example)
yn <- c("yes", "no")
a <- cpt(~asia, values = c(1, 99), levels = yn)
t.a <- cpt(~tub | asia, values = c(5, 95, 1, 99), levels = yn)
s <- cpt(~smoke, values = c(5, 5), levels = yn)
l.s <- cpt(~lung | smoke, values = c(1, 9, 1, 99), levels = yn)
b.s <- cpt(~bronc | smoke, values = c(6, 4, 3, 7), levels = yn)
e.lt <- cpt(~either | lung:tub, values = c(1, 0, 1, 0, 1, 0, 0, 1), levels = yn)
x.e <- cpt(~xray | either, values = c(98, 2, 5, 95), levels = yn)
d.be <- cpt(~dysp | bronc:either, values = c(9, 1, 7, 3, 8, 2, 1, 9), levels = yn)

# Compile the Bayesian network
chest_cpt <- compileCPT(list(a, t.a, s, l.s, b.s, e.lt, x.e, d.be))
chest_bn <- grain(chest_cpt)

# Shiny App
ui <- fluidPage(
  titlePanel("Bayesian Network: Chest Clinic (Asia Example)"),
  sidebarLayout(
    sidebarPanel(
      h3("Set Evidence"),
      uiOutput("dropdowns"),  # Dynamic dropdowns
      actionButton("apply_evidence", "Apply Evidence"),
      actionButton("retract_evidence", "Retract All Evidence")
    ),
    mainPanel(
      h3("Node Probabilities"),
      tableOutput("node_probabilities")
    )
  )
)

server <- function(input, output, session) {
  # Get node names
  node_names <- nodeNames(chest_bn)

  # Get states for each node
  node_states <- lapply(node_names, function(node) {
    states <- nodeStates(chest_bn, nodes = node)
    c("None" = "", states)  # Add "None" as the default choice
  })
  names(node_states) <- node_names

  # Generate dropdowns dynamically
  output$dropdowns <- renderUI({
    dropdowns <- lapply(node_names, function(node) {
      selectInput(
        inputId = paste0("evidence_", node),
        label = paste("Evidence for", node),
        choices = node_states[[node]],  # Use states dynamically
        selected = ""  # Default to "None"
      )
    })
    do.call(tagList, dropdowns)
  })

  # Reactive value to store the updated Bayesian network with evidence
  bn_with_evidence <- reactiveVal(chest_bn)

  # Apply evidence
  observeEvent(input$apply_evidence, {
    evidence <- lapply(node_names, function(node) {
      state <- input[[paste0("evidence_", node)]]
      if (state != "") return(c(node = node, state = state))
      NULL
    })
    evidence <- do.call(rbind, evidence)
    if (!is.null(evidence)) {
      evidence_list <- setNames(as.list(evidence[, 2]), evidence[, 1])
      updated_bn <- setEvidence(bn_with_evidence(), evidence = evidence_list)
      bn_with_evidence(updated_bn)
    }
  })

  # Retract evidence and reset dropdowns
  observeEvent(input$retract_evidence, {
    bn_with_evidence(chest_bn)  # Reset the Bayesian network

    # Reset dropdowns to "None"
    lapply(node_names, function(node) {
      updateSelectInput(session, inputId = paste0("evidence_", node), selected = "")
    })
  })

  # Display node probabilities
  output$node_probabilities <- renderTable({
    bn <- bn_with_evidence()
    prob <- querygrain(bn, nodes = node_names, simplify = TRUE)
    as.data.frame(t(prob))
  }, rownames = TRUE)
}

# Run the app
shinyApp(ui = ui, server = server)
```

**Comment by prei007** [comment]
Explanation for how evidence is applied in the code above:

Below is a more detailed explanation of what happens within the `observeEvent(input$apply_evidence, ...)` block:

1. **Triggering the Event:**  
   This code runs whenever the user clicks the "Apply Evidence" button. Shiny’s `observeEvent` detects changes in `input$apply_evidence`.

2. **Collecting User Selections:**  
   Inside the observer, `lapply(node_names, ...)` iterates over each node in the Bayesian network. For each node, it retrieves the corresponding input selection from the UI (e.g., `input$evidence_asia` for the "asia" node).

   - If the user selected an actual state (e.g., "yes" or "no") instead of the default "None," it returns a small named vector like `c(node = "asia", state = "yes")`.
   - If the user chose "None," it returns `NULL` for that node.

   After this step, `evidence` is a list of vectors, each providing a `(node, state)` pair for nodes that have chosen evidence.

3. **Filtering and Reshaping the Evidence:**  
   The line `evidence <- do.call(rbind, evidence)` stacks all these `(node, state)` pairs into a two-column structure (essentially a small data frame or matrix). If no evidence was chosen, `evidence` could be `NULL`.

4. **Preparing the Evidence for `setEvidence()`:**  
   If `evidence` is not `NULL`, we extract the node names and states. Using `setNames(as.list(evidence[, 2]), evidence[, 1])`, we create a named list where:
   
   - The list names are node names (e.g., `"asia"`),
   - The list values are the chosen states (e.g., `"yes"`).

   For example, if the user selected `asia = yes` and `smoke = no`, the resulting list would look like:  
   `list(asia = "yes", smoke = "no")`.

5. **Applying the Evidence to the Bayesian Network:**  
   With this named list, we call `setEvidence(bn_with_evidence(), evidence = evidence_list)` to update the Bayesian network’s internal representation. This incorporates the user’s chosen states into the inference, effectively conditioning all future probability queries on this evidence.

6. **Updating the Reactive Value:**  
   `bn_with_evidence(updated_bn)` sets the reactive variable to the newly conditioned Bayesian network (`updated_bn`). This ensures other parts of the app, particularly the displayed node probabilities, will now reflect the updated evidence.

In summary, the "Apply Evidence" section takes the user’s dropdown selections, organizes them into a suitable format, updates the Bayesian network with this evidence, and then refreshes the app’s state to show the new conditional probabilities.
